tools/topology/src/lib.rs

```diff
@@ -0,0 +1,450 @@
+            .currencies;
+
+        let channels = self.process_channels()?;
+
+        let host_to_dex_path =
+            Self::host_to_dex_path(&channels, &self.host_network.name, dex_network)?;
+
+        let mut currencies = vec![];
+
+        currencies.reserve_exact(dex_currencies.len());
+
+        dex_currencies
+            .iter()
+            .try_for_each(|(ticker, currency)| {
+                self.resolve_currency(dex_network, &channels, &host_to_dex_path, ticker, currency)
+                    .map(|currency| currencies.push(currency))
+            })
+            .map(|()| currencies)
+            .map_err(Into::into)
+    }
+
+    fn host_to_dex_path<'r>(
+        channels: &'r BTreeMap<&str, BTreeMap<&str, &str>>,
+        host_network: &str,
+        dex_network: &str,
+    ) -> Result<Vec<HostToDexPathChannel<'r>>, error::CurrencyDefinitions> {
+        Self::direct_host_to_dex_path(channels, host_network, dex_network).map_or_else(
+            || Self::indirect_host_to_dex_path(channels, host_network, dex_network),
+            |channel| Ok(vec![channel]),
+        )
+    }
+
+    fn direct_host_to_dex_path<'r>(
+        channels: &'r BTreeMap<&str, BTreeMap<&str, &str>>,
+        host_network: &str,
+        dex_network: &str,
+    ) -> Option<HostToDexPathChannel<'r>> {
+        channels.get(host_network).and_then(|connected_networks| {
+            connected_networks.get(dex_network).map(|&endpoint| {
+                let Some(connected_networks) = channels.get(dex_network) else {
+                    Self::unreachable_inverse_should_be_filled_in();
+                };
+
+                let Some(&inverse_endpoint) = connected_networks.get(host_network) else {
+                    Self::unreachable_inverse_should_be_filled_in();
+                };
+
+                HostToDexPathChannel {
+                    endpoint,
+                    inverse_endpoint,
+                }
+            })
+        })
+    }
+
+    fn indirect_host_to_dex_path<'r>(
+        channels: &'r BTreeMap<&str, BTreeMap<&str, &str>>,
+        host_network: &str,
+        dex_network: &str,
+    ) -> Result<Vec<HostToDexPathChannel<'r>>, error::CurrencyDefinitions> {
+        let mut endpoints_deque = Self::initial_host_to_dex_paths(channels, host_network)?;
+
+        let mut traversed_networks = BTreeSet::from([host_network]);
+
+        loop {
+            let Some((network, endpoints, mut walked_channels)) = endpoints_deque.pop_front()
+            else {
+                break Err(error::CurrencyDefinitions::HostNotConnectedToDex);
+            };
+
+            if let Some(path) = Self::explore_path_breath_first(
+                channels,
+                dex_network,
+                &mut endpoints_deque,
+                &mut traversed_networks,
+                network,
+                endpoints,
+                &mut walked_channels,
+            ) {
+                break Ok(path);
+            }
+        }
```

Comment:
I'll look into `Iterator::scan` because it rings a bell but I haven't used it or at least don't recall how it works.

Reply:
Have you looked into `scan`?

Reply:
I looked into the possibility but even in combination with `flat_map` for example, it would still have converted it to depth-first. Marking as resolved now.